autoload/lsp/ui/vim.vim

```diff
@@ -371,36 +430,201 @@ function! s:apply_workspace_edits(workspace_edits) abort
+    " command (so the import will be written on the next line) , but if we
+    " write the import first and then the \n everything will be fine.
+    " If you do not apply a command one at  time, you will need to adjust the
+    " range columns after which edit. You will get this (only one execution):
+    "
+    " execute 'keepjumps normal! 6G045laimport javax.ws.rs.Consumes;'" |
+    " execute 'keepjumps normal! 6G045la\n'
+    "
+    " resulting in this:
+    " import javax.servlet.http.HttpServletRequest;i
+    " mport javax.ws.rs.Consumes;
+    "
+    " instead of this (multiple executions):
+    " execute 'keepjumps normal! 6G045laimport javax.ws.rs.Consumes;'"
+    " execute 'keepjumps normal! 6G045li\n'
+    "
+    " resulting in this:
+    " import javax.servlet.http.HttpServletRequest;
+    " import javax.ws.rs.Consumes;
+    "
+    "
+    " The sort is also necessary since the LSP specification does not
+    " guarantee that text edits are sorted.
+    "
+    " Example:
+    " Initial text:  "abcdef"
+    " Edits:
+    " ((0,0), (0, 1), "") - remove first character 'a'
+    " ((0, 4), (0, 5), "") - remove fifth character 'e'
+    " ((0, 2), (0, 3), "") - remove third character 'c'
+    for l:text_edit in reverse(s:stable_sort(a:text_edits, function('s:compare_text_edits')))
+        let l:cmd = s:build_cmd(a:uri, l:text_edit)
+
+        try
+            let l:was_paste = &paste
+            set paste
+            execute l:cmd
+        finally
+            let &paste = l:was_paste
+        endtry
+    endfor
+
+endfunction
+
+" Sorts the elements in `array` in ascending order, but preserves the relative
+" order of the elements with equivalent values.
+"
+" `array` is a List.
+" `cmp` is a Funcref, that has the same semantic as second argument of built in
+" `sort` function.
+function! s:stable_sort(array, cmp) abort
+  let l:array_with_index = map(a:array, {idx, val -> [idx, val]})
+  let l:array_with_index = sort(l:array_with_index, {a, b -> a:cmp(a[1], b[1]) == 0 ? b[0] - a[0] : a:cmp(a[1], b[1])})
+  return map(l:array_with_index, {idx, val -> val[1]})
+endfunction
+
+" Compares two text edits, based on the starting position of the range.
+"
+" `text_edit1` and `text_edit2` are dictionaries and represent LSP TextEdit type.
+"
+" Returns 0 if both text edits starts at the same position, negative value if
+" `text_edit1` starts before `text_edit2` and positive value otherwise.
+function! s:compare_text_edits(text_edit1, text_edit2) abort
+    if a:text_edit1['range']['start']['line'] !=# a:text_edit2['range']['start']['line']
+        return a:text_edit1['range']['start']['line'] - a:text_edit2['range']['start']['line']
     endif
-    return l:result
+    if a:text_edit1['range']['start']['character'] !=# a:text_edit2['range']['start']['character']
+        return a:text_edit1['range']['start']['character'] - a:text_edit2['range']['start']['character']
+    endif
+    return 0
```

Comment:
Note there are allowed to be multiple edits at the same character where some are inserts (zero length) and exactly 0 or 1 is an overwrite. In that case the edit should be performed before the inserts.

https://github.com/natebosch/vim-lsc/blob/15f7c52c7f43802fae292fe8c61fdb95bb67ddf3/autoload/lsc/edit.vim#L185-L197

Reply:
I believe if the insert is zero length it is actually a delete operation, also the specification states this: "it is possible that multiple edits have the same start position: multiple inserts, or any number of inserts followed by a single remove or replace edit" (https://microsoft.github.io/language-server-protocol/specification#textedit)

So because we are processing in a reverse order we will perform the remove/edit first than the insert.

Reply:
I phrased that poorly, what I mean was where the replaced range was zero length.

If there is both an edit and an insert with the same start character they should not be ordered together which this will do.